cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go

```diff
@@ -18,80 +18,168 @@ package aws
 
 import (
 	"fmt"
+	"reflect"
 	"sync"
+	"time"
 
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/golang/glog"
 )
 
+const scaleToZeroSupported = false
+
 type autoScalingGroups struct {
-	registeredAsgs           []*asgInformation
-	instanceToAsg            map[AwsRef]*Asg
-	cacheMutex               sync.Mutex
-	instancesNotInManagedAsg map[AwsRef]struct{}
-	service                  autoScalingWrapper
+	registeredAsgs      []*asgInformation
+	instanceToAsg       map[AwsRef]*Asg
+	notInRegisteredAsg  map[AwsRef]bool
+	registeredAsgsMutex sync.RWMutex
+	instanceToAsgMutex  sync.Mutex
+	service             autoScalingWrapper
+	interrupt           chan struct{}
 }
 
-func newAutoScalingGroups(service autoScalingWrapper) *autoScalingGroups {
+func newAutoScalingGroups(service autoScalingWrapper) (*autoScalingGroups, error) {
 	registry := &autoScalingGroups{
-		registeredAsgs:           make([]*asgInformation, 0),
-		service:                  service,
-		instanceToAsg:            make(map[AwsRef]*Asg),
-		instancesNotInManagedAsg: make(map[AwsRef]struct{}),
+		registeredAsgs:     make([]*asgInformation, 0),
+		service:            service,
+		instanceToAsg:      make(map[AwsRef]*Asg),
+		notInRegisteredAsg: make(map[AwsRef]bool),
+		interrupt:          make(chan struct{}),
 	}
-	return registry
+	go wait.Until(func() {
+		if err := registry.regenerateCache(); err != nil {
+			glog.Errorf("Error while regenerating Asg cache: %v", err)
+		}
+	}, time.Hour, registry.interrupt)
+
+	return registry, nil
 }
 
-// Register registers asg in Aws Manager.
-func (m *autoScalingGroups) Register(asg *Asg) {
-	m.cacheMutex.Lock()
-	defer m.cacheMutex.Unlock()
+// Register ASG. Returns true if the ASG was registered.
+func (m *autoScalingGroups) Register(asg *Asg) bool {
+	m.registeredAsgsMutex.Lock()
+	defer m.registeredAsgsMutex.Unlock()
```

Comment:
I agree a single mutex would make the code cleaner, but on the surface it seems that less granular locking could result in situations where a caller is waiting for a lock unnecessarily, for example not being able to find an ASG for an instance while also registering an ASG.

That said, I mostly added a second mutex in imitation of the two-mutex pattern in the GCE manager. It's quite possible this could be premature optimisation.

Currently we use the following mutexes in the 'autoscaling group cache': 

`registeredAsgMutex`:
* Taken by a single goroutine when registering an ASG
* Taken by a single goroutine when unregistering an ASG
* Taken by one or more goroutines when listing ASGs
* Taken by one or more goroutines when regenerating the instance to ASG cache

`instanceToAsgMutex`:
* Taken by a single goroutine when finding the ASG for an instance
* Taken by a single goroutine when invalidating the instance to ASG cache
* Taken by a single goroutine when regenerating the instance to ASG cache

As far as I know there's only two goroutines competing for these mutexes, namely the one used by the RunOnce loop, and the one we spawn to regenerate the instance to ASG cache hourly. I'm new to the codebase so it's not unlikely that there's other goroutines I'm unaware of.

At the end of the day this is not something I feel strongly about. I'm curious if anyone else has an opinion?

Reply:
Note that I don't think using a single mutex would remove the need for `regenerateCacheWithoutLock`. Even with a single mutex we'd still have two cases in which we need to regenerate the cache:
1. From inside `FindForInstance`, where we have already taken a lock and thus can't take it again.
2. From everywhere else (i.e. registering an instance, unregistering an instance, regenerating the cache hourly).

Previously we dealt with this by all the 'everywhere else' cases taking the mutex explicitly. i.e. The callers would do:
```go
m.registeredAsgsMutex.Lock()
defer m.registeredAsgsMutex.Unlock()
m.regenerateCache()
```

This puts the responsibility of dealing with locking on the callers rather than the method that actually needs the lock. This is especially messy when the caller is in the AWS manager, not the autoscaling group cache struct  - it leaks implementation details of the autoscaling group cache up to the AWS manager to the point that it's almost pointless for them to be two different things.

Reply:
Thanks for clarifying again!

> for example not being able to find an ASG for an instance while also registering an ASG.
Yes - indeed that sounds problematic in larger clusters.
I'm now inclined to have fine-grained locks as you've done 👍

Reply:
I ended up reverting to a single lock after all. :) I ran into a deadlock bug in the AWS provider and after spending some time trying to debug it I was convinced that a single mutex would make life easier.